src/main/java/uk/gov/hmcts/reform/iacaseapi/infrastructure/security/idam/IdamTokenVerification.java

```diff
@@ -0,0 +1,96 @@
+package uk.gov.hmcts.reform.iacaseapi.infrastructure.security.idam;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSVerifier;
+import com.nimbusds.jose.jwk.AsymmetricJWK;
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.JWKSet;
+import com.nimbusds.jose.jwk.SecretJWK;
+import com.nimbusds.jose.proc.JWSVerifierFactory;
+import com.nimbusds.jwt.SignedJWT;
+
+import java.security.Key;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang.StringUtils;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpMethod;
+import org.springframework.web.client.RestTemplate;
+
+@Slf4j
+public class IdamTokenVerification {
+
+    private final String baseUrl;
+    private final String jwksUri;
+    private final RestTemplate restTemplate;
+    private final JWSVerifierFactory jwsVerifierFactory;
+
+    public IdamTokenVerification(String baseUrl, String jwksUri, RestTemplate restTemplate, JWSVerifierFactory jwsVerifierFactory) {
+        this.baseUrl = baseUrl;
+        this.jwksUri = jwksUri;
+        this.restTemplate = restTemplate;
+        this.jwsVerifierFactory = jwsVerifierFactory;
+    }
+
+    public boolean verifyTokenSignature(String token) {
+
+        try {
+            String tokenToCheck = StringUtils.replace(token, "Bearer ", "");
+            SignedJWT signedJwt = SignedJWT.parse(tokenToCheck);
+
+            JWKSet jsonWebKeySet = loadJsonWebKeySet();
+
+            JWSHeader jwsHeader = signedJwt.getHeader();
+            Key key = findKeyById(jsonWebKeySet, jwsHeader.getKeyID());
+
+            JWSVerifier jwsVerifier = jwsVerifierFactory.createJWSVerifier(jwsHeader, key);
+
+            return signedJwt.verify(jwsVerifier);
+        } catch (Exception e) {
+            log.error("Token validation error", e);
+
+            return false;
+        }
+    }
+
+    private JWKSet loadJsonWebKeySet() {
+
+        try {
+            return JWKSet.parse(
+                restTemplate
+                    .exchange(
+                        baseUrl + jwksUri,
+                        HttpMethod.GET,
+                        HttpEntity.EMPTY,
+                        String.class
+                    )
+                    .getBody()
+            );
+        } catch (Exception e) {
+            log.error("JWKS key loading error", e);
+            throw new RuntimeException("JWKS error", e);
```

Comment:
Could we be more specific with exceptions?

Reply:
yes, we can be more specific. I will use already provided by `auth-checker-lib` `BearerTokenInvalidException` it can be good match here.

Above implementation it taken from below page with mall changes around restTemplate:
https://tools.hmcts.net/confluence/display/SISM/OIDC+Integration+Guide

Reply:
ok, it is done